### DISCUSSION TITLE
Contact

### DISCUSSION THREAD

**Discussion by Git-Hub-Nutzer** [textbrowser/smoke#13] [Q&A] [answered]
Hi - is there a possibility to get in contact with the developer? I don't find an email anywhere...

Background:
For the german site of "Freie Messenger" I would like to add more information for Goldbug&Smoke.
At the moment everything I have is:  https://www.freie-messenger.de/sys_andere/
And in the public chat room (Freie Messenger) the echo-protocol is discussed again but nobody in there has a contact person nor forum to ask.

So if you/somebody would/could help - please contact me ...
- E-Mail: info@freie-messenger.de
- direct chat: im@jabber.de
- public chat: xmpp:freie-messenger@conference.jabber.de?join

This would be great because with more information I could add "Echo" to https://www.freie-messenger.de/systemvergleich/ .

Sorry for using the discussion function for this. Thanks a lot!

PS:
Some months ago I received the paper book "Big Seven Study" from a supporter (thanks to "unknown"). Here I have some questions too - but who could I ask? I didn't find a possibility to contact David Adams or Ann-Kathrin Maier.

**Comment by textbrowser** [comment]
Hello.

What would you like to know?

"Hello.

I'll provide some interpretations of the Echo.

The Echo is not a strict protocol. It itself does not specify a protocol. It is not contractual.

There are several implementations of the Echo: Smoke, SmokeStack, Spot-On, and Spot-On-Lite. Smoke, as an example, implements a TCP-like protocol through the Echo for file sharing. What does this mean? It means that two participants may share a file in a reliable manner without being connected directly.

[A] <-----> [B] <-----> [C] <- ... -> [X]. So, A publishes a file for X and Smoke guarantees that the transfer completes. The Steam protocol works within the Echo. It is TCP implemented through a network of applications (nodes). The [K] nodes may be Smoke, SmokeStack, Spot-On, Spot-On-Lite or something else capable of Echoing.

Spot-On-Lite introduced cryptographic routing so that specific Smoke clients receive their intended data. SO-Lite does this through inspecting special digests which Smoke clients attach to their messages. When a message is published, a non-secret digest is attached to it. This digest instructs SO-Lite nodes to direct their inbound packets to their intended destinations. If a Smoke client attaches to an SO-Lite node, it provides the node with its pseudo-identity. The SO-Lite node shares this identity through the network so that other nodes are capable of addressing messages to their correct destinations.

Spot-On introduced the concept of the Adaptive Echo. Without too much detail, this subset of the Echo allows SO nodes to be programmed with secrets so that data from a Spot-On node is only forwarded to nodes which are aware of the secrets. This arrangement creates an exclusive sub-network.

Spot-On also introduced streaming of application data through itself. A sort of private proxy. For example, it is possible to transfer a file through SO using an application which does not offer TLS by configuring two SO nodes such that both nodes are aware of the same secret. One application can write data to a local SO node while the counterpart SO node in some other network receives the ciphertext and writes the plaintext to a separate and local application.

Smoke and SmokeStack are exclusive to Android. SO and SO-Lite are available for just about any platform.

Hope this helps."

**Reply by Git-Hub-Nutzer** [reply]
First: Thanks of responding.
I / we've lot's of questions:
- who is responsible for the development of the protocol?
- is there a funding/foundation in the background?
- is there a new audit planned?
- do you know an external comparison Echo is included?
- is there a forum or public chatroom for Echo/Goldbug/Smoke/... (except freie-messenger)?
- do you know where Echo is in production or commercial use?
- are there any user statistics (yes, it's P2P!) ?
- what do you think are the most heavy differences compared to other P2P messenger systems (e.g. Briar)
- Do you know german speaking devs?
- the book "big-7" is several years old (2016) - is there an updated planned?
- do you have contact to the autors of the "big-7" or could ask them if they would contact me?
- ...

**Reply by textbrowser** [reply]
- I am. I have received some suggestions from one or two people.
- No. All of the work is voluntary and not paid.
- No.
- No.
- There is a default channel included with Smoke (Fire) and Spot-On (Buzz). Forums? Here and perhaps SourceForge.
- No.
- No. I do not track users. I was funding a public server but that became expensive.
- I don't review other systems. Echo was launched 11 years ago, so there are probably differences and similarities.
- No. I used to collaborate with a German individual but not German developers.
- You'll have to contact the mysterious authors. I was asked to review and provide edits for the book. I did not write it.
- No, I do not have their contact information.

I think the Echo is simple and splendid. I can prepare a server in a few seconds and have two devices communicating with one another in a few minutes or less (depends upon the devices). The Echo just works. I do not have to configure servers, accounts, etc. It's also quite flexible.

When I was funding the public server, my interface from home was a Raspberry Pi. Devices from home would connect to it and sometimes to a separate machine connected to the Pi. The Pi would connect to the remote server. Unfortunately, there are not many (or any) services which live on Raspberry Pis so that would have been an interesting test as well. Many scenarios were tested and these experiments are probably not mentioned anywhere. Another idea of the Echo is that it endorses exploration.

More questions?

Thanks.

**Reply by textbrowser** [reply]
The Echo is transparent. It doesn't require special software. It's naturally extensible. I think it needs imagination with implementation. :) It has four distinct and similar implementations: Smoke, SmokeStack, Spot-On, Spot-On-Lite. They all work together regardless of their differences. That is another concept of the Echo. Things that implement it should not care about other products in the network. Of course, it's not boundless.

**Reply by textbrowser** [reply]
Correction. The implementations are bounded. The Echo? No one knows.

**Reply by Git-Hub-Nutzer** [reply]
> More questions?

Sure - more than one (system, server, german contacts, offline messages, ...)

-> I'll send you an email.

**Reply by textbrowser** [reply]
Please include the questions here as this medium provides transparency.

**Reply by Git-Hub-Nutzer** [reply]
...
> Here are some more questions I/we have:
> 
> - Is there an overview how the similar implementations work / different from each other?
> 
> - Is there a server needed or not? First I thought Echo is an P2P protocol - but the website/ you say "server".
> 
> - Is "momedo" (2018) still something what I should looking for?
> 
> - Where would you place Echo in the messenger comparison https://www.freie-messenger.de/systemvergleich/ ?
>   (see the englisch version)
> 
> - How does offline messaging works?
> - Is a server needed for offline messages? On both sides?
> - Is there a simple graphic?
> 
> - As far as I know the german messenger "ginlo" (formerly SIMSme) use the same architecture like Goldbug ...
>   Do you know something about this?
>   (information is here: https://www.freie-messenger.de/warumnicht - because of closed source)
> 
> - My english is okayisch but not the best. For me it's much more easier to talk and understand in my mother language german - so if you have german contacts using Echo or I could talk to it would be very helpful for me.
...

**Reply by textbrowser** [reply]
1. No.
2. Spot-On has connecting and listening sockets. Smoke has connecting sockets. For UDP multicasting, servers are clearly not needed. I define a server as a device which issues some form of accept(). The projects do not require official servers. See demonstrations below.
3. I don't know momedo.
4. In a class of its own. :)
5. For Smoke, there are SmokeStacks and Ozones. Ozones are addresses. One or more Stacks may be aware of identical Ozones. Let's say the Ozone is Blue and let's say 10 Stacks are aware of this Ozone. In Smoke, I share my keys and it transfers them via keys generated from Blue. A friend does the same. Our messages are then recorded on those reachable Stacks. Spot-On has Institutions but the process is not so simple.
6. SmokeStacks can be both clients and servers. If they are clients, they connect to things. So no. For example, I am on Mars and my friend is on Venus. The Stacks may be on Pluto, the Sun, and on Earth. If the electrons from our devices can reach those Stacks, everything works as expected. I do not need my own personal Stack nor does my friend. We also do not need to know the upstream connection details. Think of someone configuring a bunch of Stacks and entering an Echo network.
7. Probably somewhere.
8. No. GB is a skin / theme for Spot-On.
9. I do not. Write in your preferred language and I'll use Google translate. :)

**Comment by textbrowser** [comment]
https://user-images.githubusercontent.com/10701156/128198883-905871ea-cadd-49f6-8dd4-b116ccffc4b6.mp4

https://user-images.githubusercontent.com/10701156/128199726-5afad896-f83f-408c-b557-ad063185a32b.mp4

**Comment by textbrowser** [comment]
More cats.

https://user-images.githubusercontent.com/10701156/128220477-9cac99a4-39c3-4e57-95d3-5eb567f50690.mov

netstat -ant | grep 471
tcp        0      0 192.168.178.15:4710     0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:4711            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:4712            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:4713            0.0.0.0:*               LISTEN     
tcp        0      0 192.168.178.15:4710     192.168.178.15:56130    ESTABLISHED
tcp        0      0 192.168.178.15:4710     192.168.178.15:56146    ESTABLISHED
tcp        0      0 192.168.178.15:56146    192.168.178.15:4710     ESTABLISHED
tcp        0      0 192.168.178.15:4712     192.168.178.15:53458    ESTABLISHED
tcp        0      0 192.168.178.15:46052    192.168.178.15:4713     ESTABLISHED
tcp        0      0 192.168.178.15:57582    192.168.178.15:4711     ESTABLISHED
tcp        0      0 192.168.178.15:56130    192.168.178.15:4710     ESTABLISHED
tcp        0      0 192.168.178.15:4711     192.168.178.15:57582    ESTABLISHED
tcp        0      0 192.168.178.15:4713     192.168.178.15:46052    ESTABLISHED
tcp        0      0 192.168.178.15:53458    192.168.178.15:4712     ESTABLISHED

Four listeners. All listening.

4710 has two clients, Spot-On and socat.

4711 has one client, Spot-On (57582). It’s also connected to 4712.

4712 has one client, socat (4711). It’s also connected to 4713.

4713 has one client, socat (4712). It’s also connected to 4710.

test-a ↔ 4710 ↔ 4713 ↔ 4712 ↔ 4711 ↔ test-b

Or something like that.

Two SO nodes communicating through a network of cats.

**Reply by textbrowser** [reply]
The cats are spot-less. :)

**Comment by textbrowser** [comment]
This can be achieved through a mixture of SmokeStack, Smoke (it includes non-listening Echoing), SO-Lite, SO, socat, ncat, whatever.

**Comment by textbrowser** [comment]
Perhaps Smoke will be able to also include the following scenario: non-listening (without accept()) devices connecting to each other. I was able to perform this with another project written in C++.

https://github.com/textbrowser/smoke/commit/b8d85c2d50883e9c17b52da763dd6a49f07e74cc#diff-f20b44aff67c14cb0127e62994457ef2bca6c4d0e5e3f84f0970fc8e1089b3a2

**Reply by textbrowser** [reply]
https://github.com/textbrowser/qtchess/commit/2d3013228825cb75b91d4aea4d5e84afee543790#diff-0e570ca6fabe24f94e52c1833f3ffd25567022beb826fa16891f3322051bc221

**Comment by textbrowser** [comment]
Server-less (without listeners / without accept()) peers in Spot-On.

https://github.com/textbrowser/spot-on/commit/cb8367d09156c3460505d9d1ba1a75f023bf5fe8

**Comment by textbrowser** [comment]
Server-less demonstration. STUN not required.

https://user-images.githubusercontent.com/10701156/128600620-852c9b7b-9f3b-4784-b55e-d922c467c22f.mp4

**Comment by textbrowser** [comment]
Please clarify some of the information on your site. For example, the ideas behind server-less projects. Consider Jami as an example.

https://git.jami.net/savoirfairelinux/ring-project/-/wikis/tutorials/Frequently-Asked-Questions#advanced-10

**Jami works as a server and gets new ports for each connections (randomly binded). These are the ranges that can be used for each component:

dht: UDP [4000, 8888]
audio: UDP [16384-32766]
video: UDP [49152-65534]
SIP Control: UDP/TCP randomly binded**

This documentation states that Jami works as a server. So, it is a distributed system where peers are both clients and servers. Servers exist.

What is pure P2P? The very name suggests that nothing issues accept() nor listen(). However, that is unlikely. Most implementations I've seen of P2P have nodes which behave as clients (they connect to things) and servers (they accept() and listen() for connections). The definitions are intentionally ambiguous. A peer is defined as something that cannot be connected to while a server is something that things connect to. In P2P networks, this is most likely the behavior regardless of the model (coordinated or distributed or whatever (perhaps magic cosmic dust)). P2P: clients accept() / listen() and / or connect(). Those are hybrid peers. Some are pure peers because they only connect(). As this is not my investigation, I would inquire about the specifics of the projects you list.

Above is an implementation which will be released soon demonstrating a model where two clients connect to each other without issuing accept() and listen(). They both bind to local addresses and repeatedly issue connect() every millisecond.

**Comment by textbrowser** [comment, marked as answer]
Silence.

More news! The true P2P mechanism also features TLS! It's totally optional too. Two clients can establish a TLS session by coordinating client and server modes.

Diagram.

SO Client A <- TLS -> SO Client B. Zero listeners, zero servers, infinite joy.

**Comment by textbrowser** [comment]
https://www.freie-messenger.de/sys_email/

Spot-On introduced e-mail texting long before Delta and others. It's called Poptastic and has been available since 2014.

**Comment by textbrowser** [comment]
Poptastic includes SMP, forward secrecy, key-renewal, and status information.

**Comment by textbrowser** [comment]
Released Prison Blues in February of 2025. This allows communications over GIT. Works with GPG and Spot-on and Smoke. In GitHub, private tokens allow access to repositories. A bunch of people can create some repositories and exchange private tokens. Then, they can anonymously post to one or more GIT repositories GPG messages (or other messages). Spot-On will then push and pull; GIT as a free platform for communicating. GPG attachments work.

https://github.com/textbrowser/prison-blues/commits/main/
